hera_cal/redcal.py

```diff
@@ -789,3 +837,335 @@ def is_redundantly_calibratable(antpos, bl_error_tol=1.0):
+        'g_omnical': full omnical gain dictionary (which include firstcal gains) in the same format.
+            Flagged gains will be 1.0s.
+        'gf_omnical': omnical flag dictionary in the same format. Flags arise from NaNs in log/omnical.
+        'v_omnical': omnical visibility solutions dictionary with baseline-pol tuple keys that are the
+            first elements in each of the sub-lists of reds. Flagged visibilities will be 0.0s.
+        'vf_omnical': omnical visibility flag dictionary in the same format. Flags arise from NaNs.
+        'chisq': chi^2 per degree of freedom for the omnical solution. Normalized using noise derived
+            from autocorrelations. If the inferred pol_mode from reds (see redcal.parse_pol_mode) is
+            '1pol' or '2pol', this is a dictionary mapping antenna polarization (e.g. 'Jxx') to chi^2.
+            Otherwise, there is a single chisq (because polarizations mix) and this is a numpy array.
+        'chisq_per_ant': dictionary mapping ant-pol tuples like (1,'Jxx') to the average chisq
+            for all visibilities that an antenna participates in.
+    '''    
+    assert hd.filetype == 'uvh5', 'Input HERAData object must be created with a uvh5 visibility object.'
+    
+    # get basic antenna, polarization, and observation info
+    nTimes, nFreqs = len(hd.times), len(hd.freqs)
+    antpols = list(set([ap for pol in hd.pols for ap in split_pol(pol)]))
+    ant_nums = np.unique(np.append(hd.ant_1_array, hd.ant_2_array))
+    ants = [(ant, antpol) for ant in ant_nums for antpol in antpols]
+    pol_load_list = get_pol_load_list(hd.pols, pol_mode=pol_mode)
+
+    # initialize gains to 1s, gain flags to True, and chisq to 0s
+    rv = {}  # dictionary of return values
+    rv['g_firstcal'] = {ant: np.ones((nTimes, nFreqs), dtype=np.complex64) for ant in ants}
+    rv['gf_firstcal'] = {ant: np.ones((nTimes, nFreqs), dtype=bool) for ant in ants}
+    rv['g_omnical'] = {ant: np.ones((nTimes, nFreqs), dtype=np.complex64) for ant in ants}
+    rv['gf_omnical'] = {ant: np.ones((nTimes, nFreqs), dtype=bool) for ant in ants}
+    rv['chisq'] = {antpol: np.zeros((nTimes, nFreqs), dtype=np.float32) for antpol in antpols}
+    rv['chisq_per_ant'] = {ant: np.zeros((nTimes, nFreqs), dtype=np.float32) for ant in ants}
+
+    # get reds and then intitialize omnical visibility solutions to all 1s and all flagged
+    all_reds = get_reds({ant: hd.antpos[ant] for ant in ant_nums}, pol_mode=pol_mode,
+                        pols=set([pol for pols in pol_load_list for pol in pols]))
+    all_reds = filter_reds(all_reds, ex_ants=ex_ants)
+    rv['v_omnical'] = {red[0]: np.ones((nTimes, nFreqs), dtype=np.complex64) for red in all_reds}
+    rv['vf_omnical'] = {red[0]: np.ones((nTimes, nFreqs), dtype=bool) for red in all_reds}
+
+    # solar flagging
+    lat, lon, alt = hd.telescope_location_lat_lon_alt_degrees
+    solar_alts = utils.get_sun_alt(hd.times, latitude=lat, longitude=lon)
```

Comment:
Is hd.times compatible with baseline-dependent time cadences?

Reply:
Good question. No, I don't think so. This seems like an issue for later, since it's not totally clear how we'd handle BDA data in omnical. Do we upsample? Downsample? I'm not sure. Maybe make an issue?

Reply:
Moved to issue #382.